### PR TITLE
[bug fixes] annotations <> x domains, zeros in text

### PR DESCRIPTION
**Pull request by williaster**
This PR fixes a couple bugs related to annotation usage
- when adding x-interval annotations to time series charts, the x-scale domain does not account for the annotation data. I updated the logic to account for this.
- when playing with annotations I noticed that you cannot currently enter a `0` in a `TextControl` component because the control only updates to values that evaluate `truthy`. I updated the logic to actually check for a value.

in examples below I added (`interval` or `event`) annotations with a `table slice` data source.

before (annotations clipped)
screen shot 2018-01-10 at 4 36 38 pm

after (annotations included)
screen shot 2018-01-10 at 4 36 38 pm

@mistercrunch @fabianmenges @graceguo-supercat @michellethomas 

some other bugs I noted but didn't fix:
- if the source for event or interval annotations is a _table_, passing down "from" and "until" should actually perform a filter on the table columns defined by "interval start column" (and "interval end column"), not on the slices "from" and "until" form data.
- you can't set a y-max on a chart (looked into it, tricky with nvd3 without computing your own domain; I think we should fix with data-ui integration)

### CODE REVIEW

**Comment by mistercrunch** [comment]
@fabianmenges is the best person to review this

**Comment by fabianmenges** [comment]
I'll first comment on this:

> if the source for event or interval annotations is a table, passing down "from" and "until" should actually perform a filter on the table columns defined by "interval start column" (and "interval end column"), not on the slices "from" and "until" form data.

I don't really think this is a bug but a use case that wasn't covered. The concepts of `until` and `since` are different from any other filter in Superset and the only ones I implemented. 

That said, I thought about similar use cases before and there are some foundations you can use to do what you want to do.

Take a look here:
https://github.com/apache/incubator-superset/blob/04680e5ff138f7113a3d655133307049bc91ff3d/superset/assets/javascripts/explore/components/controls/AnnotationLayer.jsx#L408

This line will override the `since` field of the annotation slice to use the `since` value of the current slice. The logic that does that is here:

https://github.com/apache/incubator-superset/blob/04680e5ff138f7113a3d655133307049bc91ff3d/superset/assets/javascripts/chart/chartAction.js#L71

Now, in order to implement what you described you will need to add something like this to the `AnnotationLayer`:
```
this.setState({ overrides: { ...overrides, filters: [
  { col: timeColumn, op: '>', val: :  'since' },
  { col: intervalEndColumn, op: '<', val: :  'until' },
] } });
```

And make changes to the `chartAction` that appends/transforms the filters correctly. This will be pretty tricky to get working in a flexible way.

File: superset/assets/visualizations/nvd3_vis.js [hunk 1]
```diff
@@ -624,13 +624,24 @@ function nvd3Vis(slice, payload) {
...
               return {
                 ...r,
-                [e.timeColumn]: timeColumn,
+                [e.timeColumn]: timeValue,
               };
-            }).filter(r => !Number.isNaN(r[e.timeColumn].getMilliseconds()));
+            }).filter((r) => {
+              const isValid = !Number.isNaN(r[e.timeColumn].getMilliseconds());
+              if (isValid) {
+                xMin = xMin < r[e.timeColumn] ? xMin : r[e.timeColumn];
```
Comment (fabianmenges) [inline superset/assets/visualizations/nvd3_vis.js]: I think its a bit weird that the filtering here has side effects. Might be cleaner to find the min and max values in that series after filtering.

File: superset/assets/visualizations/nvd3_vis.js [hunk 2]
```diff
@@ -665,16 +676,29 @@ function nvd3Vis(slice, payload) {
...
               };
-            }).filter(r => !Number.isNaN(r[e.timeColumn].getMilliseconds()) &&
-              !Number.isNaN(r[e.intervalEndColumn].getMilliseconds()));
+            }).filter((r) => {
+              const isValid = (
+                !Number.isNaN(r[e.timeColumn].getMilliseconds()) &&
+                !Number.isNaN(r[e.intervalEndColumn].getMilliseconds())
+              );
+              if (isValid) {
+                xMin = xMin < r[e.timeColumn] ? xMin : r[e.timeColumn];
```
Comment (fabianmenges) [inline superset/assets/visualizations/nvd3_vis.js]: The code currently does not really check that `timeColumn` < `intervalEndColumn` and will flip them around when necessary. Not sure if we need to handle that though.

File: superset/assets/visualizations/nvd3_vis.js (hunk 1 above)
Comment (williaster) [reply superset/assets/visualizations/nvd3_vis.js]: 👍

**Comment by mistercrunch** [comment]
@williaster let's try to get this through, please resolve conflict and merge when you have a moment

**Comment by williaster** [comment]
@mistercrunch will try to get this done today

**Comment by williaster** [comment]
I think annotations are broken in production, so will debug that as part of this.

**Comment by williaster** [comment]
fixed annotations separately in #4607, updated this PR / should be good to go so PTAL 👀

**Comment by williaster** [comment]
@mistercrunch @fabianmenges going to merge this if no objections

**Comment by graceguo-supercat** [comment]
LGTM

**Pull request merged**